src/common/upload.ts

```diff
@@ -618,4 +667,135 @@ export abstract class TurboAuthenticatedBaseUploadService
+    feeMultiplier,
+    topUpBufferMultiplier = 1.1,
+  }: {
+    totalByteCount: number;
+    maxTokenAmount?: string;
+    feeMultiplier?: number;
+    topUpBufferMultiplier?: number;
+  }): Promise<TurboCryptoFundResponse | undefined> {
+    const currentBalance = await this.paymentService.getBalance();
+    const wincPriceForOneGiB = (
+      await this.paymentService.getUploadCosts({
+        bytes: [2 ** 30],
+      })
+    )[0].winc;
+
+    const expectedWincPrice = new BigNumber(wincPriceForOneGiB)
+      .multipliedBy(totalByteCount)
+      .dividedBy(2 ** 30)
+      .toFixed(0, BigNumber.ROUND_UP);
+
+    if (
+      BigNumber(currentBalance.effectiveBalance).isGreaterThanOrEqualTo(
+        expectedWincPrice,
+      )
+    ) {
+      this.logger.debug('Sufficient balance for on demand upload', {
+        currentBalance,
+        expectedWincPrice,
+      });
+      return undefined;
+    }
+
+    this.logger.debug('Insufficient balance for on demand upload', {
+      currentBalance,
+      expectedWincPrice,
+    });
+
+    if (!this.enabledOnDemandTokens.includes(this.token)) {
+      throw new Error(
+        `On-demand uploads are not supported for token: ${this.token}`,
+      );
+    }
+
+    const topUpWincAmount = BigNumber(expectedWincPrice)
+      .minus(currentBalance.effectiveBalance)
+      .multipliedBy(topUpBufferMultiplier) // add buffer to avoid underpayment
+      .toFixed(0, BigNumber.ROUND_UP);
+
+    const wincPriceForOneToken = (
+      await this.paymentService.getWincForToken({
+        tokenAmount: tokenToBaseMap[this.token](1),
+      })
+    ).winc;
+
+    const topUpTokenAmount = new BigNumber(topUpWincAmount)
+      .dividedBy(wincPriceForOneToken)
+      .multipliedBy(tokenToBaseMap[this.token](1))
+      .toFixed(0, BigNumber.ROUND_UP);
+
+    if (maxTokenAmount !== undefined) {
+      if (new BigNumber(topUpTokenAmount).isGreaterThan(maxTokenAmount)) {
+        throw new Error(
+          `Top up token amount ${topUpTokenAmount} is greater than the maximum allowed amount of ${maxTokenAmount}`,
+        );
+      }
+    }
+
+    this.logger.debug(
+      `Topping up wallet with ${topUpTokenAmount} ${this.token} for ${topUpWincAmount} winc`,
+    );
+    const topUpResponse = await this.paymentService.topUpWithTokens({
+      tokenAmount: topUpTokenAmount,
+      feeMultiplier,
+    });
+    this.logger.debug('Top up transaction submitted', { topUpResponse });
+
+    const pollingOptions = {
+      pollIntervalMs: 3 * 1000, // poll every 3 seconds
+      timeoutMs: 60 * 1000, // wait up to 60 seconds
+    };
+
+    let tries = 0;
+    const maxTries = Math.ceil(
+      pollingOptions.timeoutMs / pollingOptions.pollIntervalMs,
+    );
+    while (topUpResponse.status !== 'confirmed' && tries < maxTries) {
+      this.logger.debug('Tx not yet confirmed, waiting to poll again', {
+        tries,
+        maxTries,
+      });
+      await sleep(pollingOptions.pollIntervalMs);
+      tries++;
```

Comment:
suggestion: move this to the `finally` block to avoid prematurely waiting on the first iteration

Reply:
the idea is that we WANT to wait the first loop because we just finished submitting the Tx immediately above